Fix the modal's responsiveness

clementprdhomme opened the pull request:
This PR fixes a couple of issues with the modal on small screens:
1. The modal would be displayed partially off-screen on mobiles
2. The content of the modal wouldn't be scrollable by default

To fix the first issue, I've imported a hook, `useMediaMatch`, from the library [rooks](https://react-hooks.org/), which allows us to detect when a media query is matched. It is used to determine if the screen is wider than 640px, Tailwind's SM breakpoint value. If the screen is narrower than that, then the horizontal translation of the content is null, otherwise -50%.

The fix is unperfect and I would appreciate some help with someone more experienced with Framer Motion: when going from a wider screen to a smaller screen, the modal is still displayed partially off-screen. The reverse is not true. In any case, if the modal is re-opened, it is displayed correctly.

The second fix makes the content of the modal scrollable by default so that mobile users can read the full content of the modal. The change is tied to a prop, `scrollable`, that can be set to `false` in case you need advanced customisation of the modal's layout (e.g. you want buttons to always be displayed at the bottom and only the centre content to be scrollable).

## Testing instructions

1. Open Storybooks and navigate to the Modal story
2. Reduce the size of the window to the one of a mobile
3. Click the button to open the modal

The modal should look ok, and the text is scrollable if it doesn't fit by default.

4. Make the window wider

The modal should still look ok and centred correctly.

andresgnlez commented on components/modal/component.tsx:
```diff
@@ -36,54 +38,101 @@ export const Modal: FC<ModalProps> = ({
   const { dialogProps } = useDialog({ 'aria-label': title }, containerRef);
 
+  // 640px corresponds to the Tailwind's sm breakpoint
+  const isSmViewport = useMediaMatch('(min-width: 640px)');
```
My only concern is what happens if a project modifies the value for `sm` breakpoint. The strong point of Tailwind is precisely changing this kind of thing, so I am a bit worried about hardcoding this value here.

clementprdhomme replied:
The problem is that it's not an easy thing with Tailwind. They recommend using babel-plugin-preval:

> we recommend using a tool like [babel-plugin-preval](https://github.com/kentcdodds/babel-plugin-preval) to generate a static version of your configuration at build-time

[Source](https://tailwindcss.com/docs/configuration#referencing-in-java-script)

I think Miguel was not a fan of this (we used it in Restor) and Next.js 12 doesn't use Babel anymore. The alternative would be storing the breakpoints in a separate JS file, and pull them in the code and in the Tailwind config file.

If that's ok with you, I could make this change.

andresgnlez commented:
I've tested the component and works really smooth 👍  Unfortunately, I have no experience with `motion` library so I will let @mbarrenechea deal with that.

clementprdhomme commented:
I've made two changes to the PR:
1. I created a new file where we can store variables that are shared between Tailwind CSS and the JS bundle, `styles.config.js`. I'm using this file to replace the previously hard-coded SM breakpoint.
2. I've fixed the issue where the modal is still be wrongly positioned when going from a wider to narrower screen. Now the modal is always correctly positioned.

mbarrenechea commented:
My only comment is related to the `rooks` library/package. For responsive we usually use [https://github.com/artsy/fresnel](https://github.com/artsy/fresnel) because is SSR friendly and as we are in a `next` app I think it fits more. https://github.com/artsy/fresnel#why-not-conditionally-render

About the animations nothing to add.

clementprdhomme commented:
@mbarrenechea I don't mind replacing `rooks` by `@artsy/fresnel`, but it seems that it doesn't entirely support variables that depend on the viewport size:

> That last point presents an interesting problem. How might we represent a component that gets styled differently at different breakpoints? (Let’s imagine a matchMedia example.) […] **We're still figuring out patterns for this**, so please [let us know](https://github.com/artsy/fresnel/issues/new) if you have suggestions.

[Source](https://github.com/artsy/fresnel#pros-vs-cons)

It seems we would have to render the content of the modal twice to use Fresnel. Is that what you expect?

mbarrenechea commented:
@clementprdhomme it's true that at server side it renders twice the content. But a modal it's not something we render at first load. Once client-side JS boots and React attaches, it simply washes over the DOM and removes markup that is unneeded, via a matchMedia call. 

So yes, I will go with [this approach](https://github.com/artsy/fresnel#pros-vs-cons)

clementprdhomme commented:
I finally had the time to update this PR. I did two changes:
1. I replaced `rooks` with `@artsy/fresnel` as Miguel requested
2. I removed a part of the modal which injected the `onDismiss` prop to the child component because it would throw an error if the child wouldn't accept the prop (like a `<div />`).

Let me know if those changes are alright.
